### PR TITLE
Modernize extension-cpp; refactor code

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87
* #88
* __->__ #86

This PR:
- creates a single unified build for extension-cpp (instead of having
  separate cpu/cuda setup.pys).
- Updates the build system to use pyproject.toml (instead of only
  setup.py)
- Uses TORCH_LIBRARY to bind operators (instead of using PyBind)

There will be more future work to add improvements (e.g. torch.compile
support) and also fix up the corresponding C++ extensions tutorial.

Test Plan:
- Refactored all of the tests under test/

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+library_name = "extension_cpp"
+
+
+def get_extensions():
+    debug_mode = os.getenv("DEBUG", "0") == "1"
+    if debug_mode:
+        print("Compiling in debug mode")
+
+    use_cuda = torch.cuda.is_available() and CUDA_HOME is not None
```
Comment (drisspg) [inline setup.py:27]: Nit also make this a env var

**Review by drisspg — APPROVED** [review]
🚀

File: README.md [hunk 2]
```diff
@@ -1,16 +1,24 @@
...
-An example of writing a C++ extension for PyTorch. See
+An example of writing a C++/CUDA extension for PyTorch. See
 [here](http://pytorch.org/tutorials/advanced/cpp_extension.html) for the accompanying tutorial.
+This repo demonstrates how to write an example `extension_cpp.ops.lltm`
+custom op that has both custom CPU and CUDA kernels.
 
-There are a few "sights" you can metaphorically visit in this repository:
+To build:
+```
+pip install .
```
Comment (albanD) [inline README.md:10]: Should we do -e to have a "develop" equivalent?

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+[build-system]
```
Comment (albanD) [inline pyproject.toml:1]: What is this file used for since we also have the setup.py ?

**Review by albanD — APPROVED** [review]
Sounds good!

**Review by bdhirsh — APPROVED** [review]

File: pyproject.toml (hunk 3 above)
Comment (zou3519) [reply pyproject.toml:1]: https://packaging.python.org/en/latest/discussions/setup-py-deprecated/#is-pyproject-toml-mandatory

"While it is not technically necessary yet, it is STRONGLY RECOMMENDED for a project to have a pyproject.toml file at the root of its source tree with a content like this:"

File: README.md (hunk 2 above)
Comment (zou3519) [reply README.md:10]: Yup

**Pull request merged**